src/sentry/utils/sentryappwebhookrequests.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import absolute_import
+
+import six
+import json
+
+from dateutil.parser import parse as parse_date
+from django.conf import settings
+from django.utils import timezone
+
+from sentry.utils import redis
+from sentry.models.sentryapp import VALID_EVENTS
+from sentry.models import Organization
+
+BUFFER_SIZE = 100
+
+
+class SentryAppWebhookRequestsBuffer(object):
+    """
+    Create a data structure to store basic information about Sentry App webhook requests in Redis
+    This should store the last 100 requests and last 100 errors (in different keys) for each event type, for each Sentry App
+    """
+
+    def __init__(self, sentry_app):
+        self.sentry_app = sentry_app
+
+        cluster_id = getattr(settings, "SENTRY_WEBHOOK_LOG_REDIS_CLUSTER", None)
+        if cluster_id is None:
+            self.client = redis.clusters.get("default").get_local_client(0)
+        else:
+            self.client = redis.redis_clusters.get(cluster_id)
+
+    def _convert_redis_request(self, redis_request, event):
+        """
+        Convert the request string stored in Redis to a python dict that can be returned from the errors endpoint
+        """
+        request = json.loads(redis_request)
+
+        formatted_request = {
+            "webhookUrl": request.get("webhook_url"),
+            "sentryAppSlug": self.sentry_app.slug,
+            "eventType": event,
+            "date": request.get("date"),
+            "responseCode": request.get("response_code"),
+        }
+
+        if "organization_id" in request:
+            org = Organization.objects.get(id=request["organization_id"])
```

Comment:
@nolachen What happens if we don't find a matching org?

Reply:
@scefali that would be an error but the org id should always be passed in when adding a request to the buffer. should this be handled more?

Reply:
@nolachen Is it a 500 error or a 4xx error?

Reply:
@scefali probably a 500, all the requests are added without user input anyway

Reply:
@nolachen we should try to make sure we don't return a 500 if we don't get an org match. You could use `catch_raised_errors` and add a case for `Organization.DoesNotExist`.

Reply:
@scefali wait this isn't an API endpoint though, it gets called eventually from the endpoint but there's nothing the user could do to actually fix it if the org doesn't exist so I don't think its a 400. If the org didn't exist it's because a bad org_id was passed into the `add_request` function which only gets called in the backend

Reply:
@nolachen If you raise an error, it should bubble up to the top level. And you are right it shouldn't be a 400, it should be a 404 because you can't find the org. Does that make sense?

Reply:
@scefali  hmm i guess i'm still a little confused why it should be a 404 since it would be the fault of whoever called the `add_request` function instead of anything to do with the user

Reply:
Also,

`Organization.objects.get_from_cache` please.

From my perspective, a 500 on stuff like this is weird. It's an error, sure. But for a 500, who does it notify that its' an error? Us. And it's not something that's actionable to us. We're jsut going to log it into Sentry and ignore it. The only better argument I'd see is if we catch the error, handle it, then explicitly return a 500 so that we don't alert ourselves.

Providing some 4xx code suggests to the user it's their mistake in the request data, and we get a better chance to return back a helpful error message with it.